File: src/clashroyale/entities/Building.java
package clashroyale.entities;

import clashroyale.core.GameEngine;
import clashroyale.models.Card;

public class Building extends Entity {
    public Card c;
    public int cd;
    public Entity lk;
    public boolean atk;

    public Building(int t, double x, double y, Card c) {
        super(t, x, y, c.hp, c.n.equals("Cannon") ? 15 : 20, 10000, false, false);
        this.c = c;
    }

    @Override
    public void act(GameEngine g) {
        if (fr > 0) { infernoTick = 0; return; }
        if (st-- > 0) { infernoTick = 0; return; }

        // Lifetime decay
        hp -= mhp / c.ms; // Use c.ms (lifetime in frames)
        if (hp <= 0) {
            if (c.n.equals("Elixir Collector")) {
                g.giveElixir(tm, 1.0); // Give 1 elixir on death
            }
            return;
        }

        // Elixir Collector Logic
        if (c.n.equals("Elixir Collector")) {
            // 13 seconds = 13 * 60 = 780 frames
            infernoTick++;
            if (infernoTick >= 780) {
                g.giveElixir(tm, 1.0);
                infernoTick = 0;
            }
            return; // Pump doesn't attack
        }

        if (lk != null) {
            if (lk.hp <= 0 || dist(lk) > c.si) { lk = null; atk = false; infernoTick = 0; }
        }

        if (!atk) {
            double min = 9999;
            Entity best = null;
            for (Entity e : g.ents) {
                if (e.tm != tm) {
                    if (e.fly && !c.ar) continue; // Cannot hit air if not air-targeting
                    double d = dist(e);
                    if (d < min && d < c.si) { min = d; best = e; }
                }
            }
            if (best != null) lk = best;
        }

        double myHitbox = g.getHitboxRadius(this);
        double targetHitbox = (lk != null) ? g.getHitboxRadius(lk) : 0;
        double attackRange = c.rn + myHitbox + targetHitbox + 2;

        if (lk != null && dist(lk) <= attackRange) {
            atk = true;
            if (cd-- > 0) return;
            
            // Attack
            if (c.n.equals("Inferno Tower")) {
                infernoTick++;
                int stage = infernoTick / 25;
                int mult = getInfernoMultiplier(stage);
                
                int dmg = c.d * mult;
                lk.hp -= dmg;
            } else {
                g.projs.add(new Proj(x, y, lk.x, lk.y, lk, 8, false, 4, c.d, tm, false));
            }
            cd = c.rt;
        } else {
            atk = false;
            infernoTick = 0;
            cd = c.rt;
        }
    }
}

File: web/src/entities/Building.js
import Entity from './Entity.js';
import Proj from './Proj.js';
// Troop import removed to avoid circular dependency

export default class Building extends Entity {
    constructor(t, x, y, c) {
        let radius = (c.n === "Cannon") ? 15 : 20;
        if (c.n === "Crate") radius = 14;
        super(0, t, x, y, c.hp, radius, 10000, false, false);
        this.c = c;
        this.cd = 0;
        this.lk = null;
        this.atk = false;
    }

    act(g) {
        if (this.fr > 0) { this.infernoTick = 0; return; }
        if (this.st-- > 0) { this.infernoTick = 0; return; }

        // Lifetime decay
        this.hp -= this.mhp / this.c.ms;
        if (this.hp <= 0) {
            // Death Logic handled via die(g) called by GameEngine
            return;
        }

        // Elixir Collector Logic
        if (this.c.n === "Elixir Collector") {
            this.infernoTick++;
            if (this.infernoTick >= 780) {
                g.giveElixir(this.tm, 1.0);
                this.infernoTick = 0;
            }
            return;
        }

        if (this.lk) {
            if (this.lk.hp <= 0 || this.dist(this.lk) > this.c.si) {
                this.lk = null;
                this.atk = false;
                this.infernoTick = 0;
            }
        }

        if (!this.atk) {
            let min = 9999;
            let best = null;
            for (let e of g.ents) {
                if (e.tm !== this.tm) {
                    if (e.fly && !this.c.ar) continue;
                    let d = this.dist(e);
                    if (d < min && d < this.c.si) {
                        min = d;
                        best = e;
                    }
                }
            }
            if (best) this.lk = best;
        }

        let myHitbox = g.getHitboxRadius(this);
        let targetHitbox = (this.lk) ? g.getHitboxRadius(this.lk) : 0;
        let attackRange = this.c.rn + myHitbox + targetHitbox + 2;

        if (this.lk && this.dist(this.lk) <= attackRange) {
            this.atk = true;
            if (this.cd-- > 0) return;

            // Attack
            if (this.c.n === "Inferno Tower") {
                this.infernoTick++;
                let stage = Math.floor(this.infernoTick / 25);
                let mult = this.getInfernoMultiplier(stage);
                let dmg = this.c.d * mult;
                this.lk.hp -= dmg;
            } else {
                g.projs.push(new Proj(this.x, this.y, this.lk.x, this.lk.y, this.lk, 8, false, 4, this.c.d, this.tm, false));
            }
            this.cd = this.c.rt;
        } else {
            this.atk = false;
            this.infernoTick = 0;
            this.cd = this.c.rt;
        }
    }

    die(g) {
        if (this.c.n === "Elixir Collector") {
            g.giveElixir(this.tm, 1.0);
        } else if (this.c.n === "Crate") {
            g.handleCrateDeath(this);
        }
    }
}
